address/address.go

```diff
@@ -47,327 +77,236 @@ const (
-		}
+	log.Debugf("new address protocol: %x, payload: %v", protocol, payload)
+	return Address{string(buf)}
+}
+
+func encode(network Network, addr Address) (string, error) {
+	// TODO(frrist): not sure if this is the correct thing to do here
+	if addr == Undef {
+		return "", nil
 	}
-	if version > 31 {
-		return "", fmt.Errorf("version too large")
+	var ntwk string
+	switch network {
+	case Mainnet:
+		ntwk = MainnetPrefix
+	case Testnet:
+		ntwk = TestnetPrefix
+	default:
+		panic("invalid network byte")
 	}
 
-	hrp = strings.ToLower(hrp)
-	combined := append([]byte{version}, Base32.EncodeToBytes(data)...)
-	sum := createChecksum(hrp, combined)
-
-	toEncode := append(combined, sum...)
-	encoded := make([]byte, len(toEncode))
-
-	for i, p := range toEncode {
-		if p >= 32 {
-			return "", fmt.Errorf("invalid data")
-		}
-		encoded[i] += Base32Charset[p]
+	var strAddr string
+	switch addr.Protocol() {
+	case SECP256K1, Actor, BLS:
+		cksm := Checksum(append([]byte{addr.Protocol()}, addr.Payload()...))
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + AddressEncoding.WithPadding(-1).EncodeToString(append(addr.Payload(), cksm[:]...))
+	case ID:
+		strAddr = ntwk + fmt.Sprintf("%d", addr.Protocol()) + fmt.Sprintf("%d", leb128.ToUInt64(addr.Payload()))
+	default:
+		return "", errors.New("invalid protocol byte")
 	}
-
-	return hrp + string(encoded), nil
+	log.Debugf("encoded address: %s", strAddr)
+	return strAddr, nil
 }
 
-// decode decodes a filecoin address and returns
-// the hrp(human-readable part), version (1byte) and data(32bit data array)
-// or an error.
-func decode(addr string) (string, byte, []byte, error) {
-	if len(addr) < 2 {
-		return "", 0, nil, fmt.Errorf("address too short")
-	}
-	if len(addr) > 41 {
-		return "", 0, nil, fmt.Errorf("too long: len=%d", len(addr))
+func decode(a string) (Address, error) {
+	if len(a) == 0 {
+		return Undef, nil
 	}
-	if strings.ToLower(addr) != addr && strings.ToUpper(addr) != addr {
-		return "", 0, nil, fmt.Errorf("mixed case")
+	log.Debugf("decoding address: %s", a)
+	if len(a) < 3 {
+		return Undef, errors.Errorf("invalid address: %s length: %d, too short", a, len(a))
 	}
-	addr = strings.ToLower(addr)
-	hrp := addr[0:2]
-	for p, c := range hrp {
-		if c < 33 || c > 126 {
-			return "", 0, nil, fmt.Errorf("invalid character human-readable part: addr[%d]=%d", p, c)
-		}
-	}
-
-	// check data
-	data := addr[2:]
-	dataBytes := []byte{}
-	for _, b := range data {
-		// alphanumeric only
-		if !((b >= int32('0') && b <= int32('9')) || (b >= int32('A') && b <= int32('Z')) || (b >= int32('a') && b <= int32('z'))) {
-			return "", 0, nil, fmt.Errorf("non alphanumeric character")
-		}
 
-		// exclude [1,b,i,o]
-		if b == int32('1') || b == int32('b') || b == int32('i') || b == int32('o') {
-			return "", 0, nil, fmt.Errorf("invalid character")
-		}
+	if string(a[0]) != MainnetPrefix && string(a[0]) != TestnetPrefix {
```

Comment:
I don't like the way we're half supporting networks. 

1. Testnet and Mainnet addresses decode to the same internal representation of an address.
2. We have no way of validating that an address encoded with a network can only be used in that network.
3. All addresses are encoded to Mainnet addresses.

This PR could: 

1. Add proper support for networks meaning that the node would be configured somehow with the right network (ideally in the genesis block somehow), use that prefix for encoding, and only permit that prefix in decoding. Also consider the implications of transmitting cbor encoded addresses that drop network information.
2. Table the above and require (and assume) all addresses to be Testnet addresses, since Testnet is our first target.
3. Table the concept of encoding the network at all. All addresses begin with 'f'. This makes it easier to distinguish our addresses from other cryptocurrencies. It provides no information otherwise.

@whyrusleeping ?

Reply:
option 2 seems like the right approach if we want to keep the distinction. option 3 is what i would personally push for. I still don't think the difference between testnet and mainnet addressses makes that much of a difference.

Reply:
I have opened an issue to track this question here: https://github.com/filecoin-project/go-filecoin/issues/2220. For now I will go with option 2 in this PR.